### PR TITLE
Get RUN_AS_DEVIN working with app sandbox

### PR DESCRIPTION
**Pull request by xingyaoww**
The reason why `RUN_AS_DEVIN` was not working properly is that `SANDBOX_USER_ID` is not properly setted up. `os.getpid()` from the `app` container will return `UID=0`, which creates the issue of https://github.com/OpenDevin/OpenDevin/issues/936.

This PR fix:
*  Allowing `RUN_AS_DEVIN` with app container. The caveat is that the user need to pass-in their UID when starting the `app` container. I have updated the readme accordingly
```
docker run \
    -e LLM_API_KEY \
    -e WORKSPACE_MOUNT_PATH=$WORKSPACE_BASE \
    -v $WORKSPACE_BASE:/opt/workspace_base \
    -e SANDBOX_USER_ID=$(id -u) \
    -v /var/run/docker.sock:/var/run/docker.sock \
    -p 4000:3000 \
    --add-host host.docker.internal=host-gateway \
    ghcr.io/opendevin/opendevin:latest
```

* Make `SANDBOX_USER_ID` consistent: the `app` will run with a user with UID of `SANDBOX_USER_ID,` which won't cause a permission issue in writing to `workspace`; the `sandbox` will run with a UID of `SANDBOX_USER_ID` as well.

### CODE REVIEW

**Comment by li-boxuan** [comment]
Seems integration tests have captured a bug https://github.com/OpenDevin/OpenDevin/actions/runs/8866160837/job/24343257827?pr=1426#step:6:66:

> chmod: changing permissions of 'hello.sh': Operation not permitted

**Comment by xingyaoww** [comment]
Good catch! Seems like a mounting issue, already pushed a fix see if we can pass that
https://github.com/moby/moby/issues/2259#issuecomment-51535443

File: containers/app/Dockerfile [hunk 1]
```diff
@@ -32,21 +32,32 @@ FROM python:3.12-slim as runtime
 
 WORKDIR /app
 
-ENV RUN_AS_DEVIN=false
-ENV USE_HOST_NETWORK=false
+ENV RUN_AS_DEVIN=true
+ENV SANDBOX_USER_ID=1000
+ENV USE_HOST_NETWORK=true
```
Comment (rbren) [inline containers/app/Dockerfile:37]: I still think we need to leave host network off until it's out of beta for Docker Desktop users. Is this a critical part of this PR?

File: containers/app/entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  echo "SANDBOX_USER_ID is not set"
+  exit 1
+fi
+
+# change uid of opendevin user to match the host user
+# but the group id is not changed, so the user can still access everything under /app
+usermod -u $SANDBOX_USER_ID opendevin
+
+# make docker.sock accessible to the user
+chmod 777 /var/run/docker.sock
```
Comment (rbren) [inline containers/app/entrypoint.sh:18]: This seems dangerous--IIUC it's going to change permissions _on the host_, because we mount it

**Review by rbren — CHANGES_REQUESTED** [review]
I think I get why this is necessary--we have to plumb the user's ID into _both_ the app container and the sandbox to keep file ownership/permissions consistent. It's a lot to manage but I can't think of a better solution :/

The networking stuff here is what concerns me most though--I definitely don't think we should be changing permissions on docker.sock

File: containers/app/Dockerfile (hunk 1 above)
Comment (xingyaoww) [reply containers/app/Dockerfile]: reverted!

File: containers/app/entrypoint.sh (hunk 2 above)
Comment (xingyaoww) [reply containers/app/entrypoint.sh]: fixed in the latest commit by changing the `opendevin` user's gid!

File: .github/workflows/ghcr.yml [hunk 3]
```diff
@@ -42,8 +42,21 @@ jobs:
           username: ${{ github.repository_owner }}
           password: ${{ secrets.GITHUB_TOKEN }}
 
-      - name: Delete huge unnecessary tools folder
-        run: rm -rf /opt/hostedtoolcache
+      - name: Free Disk Space (Ubuntu)
```
Comment (rbren) [inline .github/workflows/ghcr.yml:45]: we really need to fix the underlying issue here 😅

**Review by rbren — APPROVED** [review]

**Comment by rbren** [comment]
Thanks for doing this!

**Pull request merged**